lib/tokens.js

```diff
@@ -0,0 +1,687 @@
+        [`tokenBatch.resolution.${encodedRequester}`]: entry
+      }
+    }, database.writeOptions);
+    if(result.result.n === 0) {
+      // another process resolved a token concurrently, try again
+      continue;
+    }
+
+    // return new pairwise token
+    return {pairwiseToken};
+  }
+}
+
+/**
+ * Resolves a token to the `internalId` to which it is linked. This
+ * `internalId` should never be shared outside of the system that uses this
+ * module. This function is useful for obtaining the `internalId` associated
+ * with a token such that registered, encrypted documents associated with the
+ * `internalId` can be retrieved.
+ *
+ * @param {object} options - Options to use.
+ * @param {string} options.token - The token to resolve.
+ *
+ * @returns {object} An object with the `internalId`.
+ */
+export async function resolveToInternalId({token}) {
+  // parse token
+  const {batchId} = await _parseToken({token});
+
+  // get batch document
+  const {tokenBatch} = await _getBatch({id: batchId});
+  const {internalId} = tokenBatch;
+
+  // special case resolve to internal ID with flag set
+  return {internalId};
+}
+
+async function _createToken({
+  hmac, tokenVersion, tokenBatch, index, attributes
+} = {}) {
+  // decode batch ID
+  const batchId = base64url.decode(tokenBatch.id);
+
+  // get version options
+  const {id: version, options: {batchSaltSize}} = tokenVersion;
+
+  // build data to encrypt/wrap: batchId|index|aadSize|aad|padding
+  // total = 192-bits, 24 bytes
+  const toWrap = new Uint8Array(UNWRAPPED_SIZE);
+  let dv = new DataView(toWrap.buffer, toWrap.byteOffset, toWrap.length);
+  let offset = 0;
+  // batch ID, `batchIdSize`, default is 16 bytes
+  toWrap.set(batchId, offset);
+  // index, 1 byte
+  toWrap[offset += batchId.length] = index;
+  // aadSize, 1 byte
+  toWrap[offset += 1] = attributes.length;
+  // attributes (up to 8 bytes max, but affects token size)
+  toWrap.set(attributes, offset += 1);
+  // random padding for remaining bytes
+  offset += attributes.length;
+  const padding = await randomBytesAsync(toWrap.length - offset);
+  toWrap.set(padding, offset);
+
+  // generate salt based on token version options
+  const salt = await randomBytesAsync(batchSaltSize);
+
+  // create KEK via HMAC(version|salt)
+  const kek = await _getKek({hmac, version, salt});
+
+  // wrap `toWrap` as if it were a key, output is 8 more bytes than input
+  const wrapped = await kek.wrapKey({unwrappedKey: toWrap});
+
+  // build token: version|salt|wrapped|attributes(aad)
+  // example `tokenSize`:
+  // 2 bytes: version
+  // 16 bytes: salt (larger reduces key collisions, but increases token size)
+  // 32 bytes: wrapped data
+  // 8 bytes: attributes (max of 8, can be 0 to reduce token by 8 bytes)
+  // = 58 byte token max, 50 bytes minimum
+  const tokenSize = 2 + salt.length + wrapped.length + attributes.length;
+  const token = new Uint8Array(tokenSize);
```

Comment:
- related to the above, if the wrapped size is not a constant 24+8, then the comment here should be adjusted
- the salt size is a param, so the comment sizes could be off if that's changed
- array size dynamic so that should be ok
- are the attributes supposed to be added both wrapped and here?

Reply:
Yes, this will become more dynamic and yes the attributes are supposed to be both wrapped and duplicated here. The wrapping provides authentication for them and the duplication in the clear allows them to be read by external systems.